contrib/src/main/scala/akka/stream/contrib/SourceGen.scala

```diff
@@ -7,27 +7,35 @@ import akka.stream._
 import akka.stream.scaladsl._
 import akka.stream.stage._
 import com.typesafe.config.ConfigFactory
-import scala.concurrent.duration.Duration
+import scala.concurrent.duration.{ Duration, FiniteDuration }
 
 /**
  * Source factory methods are placed here
  */
 object SourceGen {
+  private[this] lazy val defaultTimeout = Duration.fromNanos(ConfigFactory.load().getDuration("akka.stream.contrib.unfold-flow-timeout").toNanos)
 
   /**
    * Create a `Source` that will unfold a value of type `S` by
    * passing it through a flow. The flow should emit a
    * pair of the next state `S` and output elements of type `E`.
    * Source completes when the flow completes.
    *
+   * The `timeout` parameter specifies waiting time after inner
+   * flow provided by the user for unfold flow API cancels
+   * upstream, to get also the downstream cancelation (as
+   * graceful completion or failure which is propagated).
+   * If inner flow fails to complete/fail downstream, stage is
+   * failed with an IllegalStateException.
+   *
    * IMPORTANT CAVEAT:
    * The given flow must not change the number of elements passing through it (i.e. it should output
    * exactly one element for every received element). Ignoring this, will have an unpredicted result,
    * and may result in a deadlock.
    */
-  def unfoldFlow[S, E, M](seed: S)(flow: Graph[FlowShape[S, (S, E)], M]): Source[E, M] = {
+  def unfoldFlow[S, E, M](seed: S, timeout: FiniteDuration)(flow: Graph[FlowShape[S, (S, E)], M]): Source[E, M] = {
```

Comment:
To make this possible and convenient at least for the Scala API we could make this timeout parameter implicit. Also it would be good to use the `Timeout` type here.